### PR TITLE
Reserve Withdrawals

### PR DESCRIPTION
**Pull request by migueldingli1997**
**New messages**:
- `MsgWithdrawReserve`: withdraws tokens from the reserve (available both from CLI and REST). Only allowed if the withdrawer is the bond's controller DID, and if the bond is in the `OPEN` state and has enough available reserve.

**New fields in `bond` type**:
- `ReserveWithdrawalAddress`: destination address of any reserve tokens withdrawn via the `MsgWithdrawReserve` message. This is specified during bond creation and is not optional.
- `AvailableReserve`: coins value specifying the remaining reserve out of the current (virtual) reserve. This will only ever not match with the current reserve if the `MsgWithdrawReserve` message is used. This defaults to `nil` (empty coins) and is _not_ specified during bond creation.
- `AllowReserveWithdrawals`: bool specifying whether reserve withdrawals by the controller DID are allowed. This is specified  during bond creation and is not optional (however if excluded from the CLI, it is assumed to be False).

**New queries**:
- `QueryAvailableReserve`: gets the bond's available reserve (available both from CLI and REST)

**New invariants**:
- `AvailableReserveInvariant`: ensures that the sum of all the 'available reserve' values being reported by the bonds matches the balance in the global reserve pool.

**Other changes**:
- Updated documentation and scripts/demos
- Renamed `WithdrawReserve` function to `WithdrawFromReserve` to avoid clash with new message
- Renamed `DepositReserve` function to `DepositIntoReserve` to match above change

**Interesting points**:
- Wherever we use `setReserveBalances`, we now also update the available reserve via `setAvailableReserve`
- Sells are not allowed when reserve withdrawals are enabled (so the two bools cannot be True in the bond)

### CODE REVIEW

File: x/bonds/client/rest/query.go [hunk 1]
```diff
@@ -141,6 +142,22 @@ func queryCurrentReserveHandler(clientCtx client.Context) http.HandlerFunc {
...
 }
 
+func queryAvailableReserveHandler(clientCtx client.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		bondDid := vars[RestBondDid]
+
+		res, _, err := clientCtx.QueryWithData(
+			fmt.Sprintf("custom/%s/%s/%s",
+				types.QuerierRoute, keeper.QueryCurrentReserve, bondDid), nil)
```
Comment (stefaniatadama) [inline x/bonds/client/rest/query.go:152]: I think this should be `keeper.QueryAvailableReserve`.

File: x/bonds/client/rest/tx.go [hunk 2]
```diff
@@ -444,3 +467,36 @@ func withdrawShareRequestHandler(clientCtx client.Context) http.HandlerFunc {
 		tx.WriteGeneratedTxResponse(clientCtx, w, req.BaseReq, msg)
 	}
 }
+
+type withdrawReserveReq struct {
+	BaseReq      rest.BaseReq `json:"base_req" yaml:"base_req"`
+	BondDid      string       `json:"bond_did" yaml:"bond_did"`
+	Amount       string       `json:"amount" yaml:"amount"`
+	RecipientDid string       `json:"recipient_did" yaml:"recipient_did"`
```
Comment (stefaniatadama) [inline x/bonds/client/rest/tx.go:475]: (Improvement) Maybe this should be renamed to `WithdrawerDid` for consistency?

File: x/bonds/keeper/msg_server.go [hunk 3]
```diff
@@ -747,3 +759,67 @@ func (k msgServer) WithdrawShare(goCtx context.Context, msg *types.MsgWithdrawSh
...
+	}
+
+	if !bond.AllowReserveWithdrawals {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized,
+			"bond does not allow reserve withdrawals")
+	}
+
+	if bond.ControllerDid != msg.WithdrawerDid {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized,
+			"recipient must be the controller of the bond")
```
Comment (stefaniatadama) [inline x/bonds/keeper/msg_server.go:791]: (Improvement) withdrawer?

**Review by ferendo — COMMENTED** [review]
Looks good to me

File: x/bonds/client/rest/query.go (hunk 1 above)
Comment (migueldingli1997) [reply x/bonds/client/rest/query.go]: Thanks

File: x/bonds/client/rest/tx.go (hunk 2 above)
Comment (migueldingli1997) [reply x/bonds/client/rest/tx.go:475]: Definitely

File: x/bonds/keeper/msg_server.go (hunk 3 above)
Comment (migueldingli1997) [reply x/bonds/keeper/msg_server.go]: Thanks

File: x/bonds/spec/01_concepts.md [hunk 4]
```diff
@@ -51,15 +51,18 @@ type Bond struct {
     ReserveTokens                []string                                 
     TxFeePercentage              sdk.Dec   
     ExitFeePercentage            sdk.Dec   
-    FeeAddress                   string                                   
+    FeeAddress                   string
+	ReserveWithdrawalAddress     string
```
Comment (stefaniatadama) [inline x/bonds/spec/01_concepts.md:55]: Fix indent.

Comment (migueldingli1997) [reply x/bonds/spec/01_concepts.md]: Fixed

**Review by stefaniatadama — APPROVED** [review]

**Pull request merged**